ext/msgpack/unpacker.c

```diff
@@ -678,7 +678,7 @@ int msgpack_unpacker_read_array_header(msgpack_unpacker_t* uk, uint32_t* result_
     } else if(b == 0xdc) {
         /* array 16 */
         READ_CAST_BLOCK_OR_RETURN_EOF(cb, uk, 2);
-        *result_size = _msgpack_be16(cb->u16);
+        *result_size = _msgpack_be16(cb->u16) & 0x0000FFFF; /* expand value to 32 bits */
```

Comment:
Shouldn't that be inside `_msgpack_be16`?

Reply:
Hi @byroot - you're right. It should be there and I had assumed that it must have been working fine since it's also should been used in multiple places. It seems not.

I will raise a PR with the change as I noted in:
https://github.com/msgpack/msgpack-ruby/pull/297#issuecomment-1224531953

I didn't want to do it at that time of the night.